llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -690,6 +690,13 @@ void SelectionDAGLegalize::LegalizeLoadOps(SDNode *Node) {
       assert(NVT.getSizeInBits() == VT.getSizeInBits() &&
              "Can only promote loads to same size type");
 
+      // If the range metadata type does not match the legalized memory
+      // operation type, remove the range metadata.
+      if (const MDNode *MD = LD->getRanges()) {
+        ConstantInt *Lower = mdconst::extract<ConstantInt>(MD->getOperand(0));
+        if (Lower->getBitWidth() != NVT.getScalarSizeInBits())
+          LD->getMemOperand()->clearRanges();
+      }
```

Comment:
In principle we should cast the range. But since this references IR metadata directly, this may imply mutating the IR which we should avoid doing

Reply:
Is there a way to attach the range metadata to the subsequent bitcast to the original type?